File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/job/meta/DownloadJobRequest.scala
package com.webank.eggroll.core.deepspeed.job.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.containers.meta.ContentType
import com.webank.eggroll.core.meta.Deepspeed

import scala.collection.JavaConverters._
import scala.language.implicitConversions

case class DownloadJobRequest(sessionId: String, ranks: Array[Int],
                              compressMethod: String,
                              compressLevel: Int = 1,
                              contentType: ContentType.ContentType)

object DownloadJobRequest {
  implicit def serialize(src: DownloadJobRequest): Array[Byte] = {
    Deepspeed.DownloadJobRequest.newBuilder()
      .setSessionId(src.sessionId)
      .addAllRanks(src.ranks.map(_.asInstanceOf[java.lang.Integer]).toSeq.asJava)
      .setCompressMethod(src.compressMethod)
      .setCompressLevel(src.compressLevel)
      .build()
      .toByteArray
  }

  implicit def deserialize(byteString: ByteString): DownloadJobRequest = {
    val proto = Deepspeed.DownloadJobRequest.parseFrom(byteString)
    DownloadJobRequest(
      sessionId = proto.getSessionId,
      ranks = proto.getRanksList.asScala.toArray.map(_.toInt),
      compressMethod = proto.getCompressMethod,
      compressLevel = proto.getCompressLevel,
      contentType = proto.getContentType
    )
  }
}

File: jvm/core/src/main/java/org/fedai/eggroll/core/pojo/DownloadJobRequest.java
package org.fedai.eggroll.core.pojo;

import com.webank.eggroll.core.meta.Containers;
import com.webank.eggroll.core.meta.Deepspeed;
import com.webank.eggroll.core.meta.Meta;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

@Data
public class DownloadJobRequest implements RpcMessage {

    Logger log = LoggerFactory.getLogger(DownloadJobRequest.class);

    private String sessionId;
    List<Integer> ranks;
    private String compressMethod;
    private Integer compressLevel = 1;
    Containers.ContentType contentType;


    public Deepspeed.DownloadJobRequest toProto() {
        Deepspeed.DownloadJobRequest.Builder builder = Deepspeed.DownloadJobRequest.newBuilder();
        builder.setSessionId(this.sessionId);
        builder.setCompressMethod(this.compressMethod);
        builder.addAllRanks(this.ranks);
        builder.setCompressLevel(this.compressLevel);
        builder.setContentType(this.contentType);

        return builder.build();
    }

    @Override
    public byte[] serialize() {
        return toProto().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Deepspeed.DownloadJobRequest downloadJobRequest = Deepspeed.DownloadJobRequest.parseFrom(data);
            this.sessionId = downloadJobRequest.getSessionId();
            this.ranks = downloadJobRequest.getRanksList();
            this.compressMethod = downloadJobRequest.getCompressMethod();
            this.compressLevel = downloadJobRequest.getCompressLevel();
            this.contentType = downloadJobRequest.getContentType();
        } catch (Exception e) {
            log.error("deserialize error : ", e);
        }
    }
}
